katcp/test/test_client.py

```diff
@@ -310,6 +315,182 @@ def notify_connected(connected):
                         "Expected %r to not be %r" % (stream, self.client._stream))
         self.assertEqual(sockname, self.client._stream.socket.getpeername())
 
+
+class TestDeviceClientMemoryLeaks(tornado.testing.AsyncTestCase):
+
+    def setUp(self):
+        super(TestDeviceClientMemoryLeaks, self).setUp()
+        self.server = DeviceTestServer('', 0)
+        start_thread_with_cleanup(self, self.server, start_timeout=0.1)
+        self.host, self.port = self.server.bind_address
+
+    def test_no_memory_leak_managed_ioloop(self):
+        client = katcp.DeviceClient(self.host, self.port)
+        wr = weakref.ref(client)
+
+        client.start(timeout=0.1)
+        client.wait_protocol(timeout=0.1)
+        self.assertTrue(client.protocol_flags)
+        client.stop(timeout=0.1)
+        client.join(timeout=0.1)
+
+        # clear strong reference and check if object can be garbage collected
+        client = None
+        gc.collect()
```

Comment:
I don't know if katcp aims to support pypy at all, but my experience has been that pypy sometimes needs multiple calls to `gc.collect()` to fully collect all unreachable objects (it cleans out objects that have no references, but doesn't recurse to objects that those collected objects were referencing).

Reply:
Interesting.  I've never thought about using katcp in pypy  - no plans to support it.  We can update the tests, if that ever happens.